### ISSUE TITLE
Missing nullable reference types' warnings on invalid overrides

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#71477]
**Version Used**: 
```
Branch main (20 Dec 2023)
Latest commit [0caa9d1](https://github.com/dotnet/roslyn/commit/0caa9d162683c44645ccc6e1645ce4a01acfd8dc) by Cyrus Najmabadi:
Merge pull request #71367 from CyrusNajmabadi/collectionExprInInit
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

A b = new B();
var s = b.M<string?>(null);
System.Console.WriteLine(s.Length);

public interface I { }
public abstract class A
{
    [return: NotNull] public abstract T M<T>(T t) where T : class?;
}

class B : A
{
    public override T M<T>(T t) where T : class => t;
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQAAgAwl0CMAdACJQQDmMA9gGcALlADGgtgGF+CIgEEKOAJ6CoggNy1a8ksBIBeEjCIB3EgCEAFAEotNAG4QATiUGG9bALIAeVgwB+AD4rbDw7WlY2VgBOK0kAGXIeYQALCJpIgGYSWGEiZwAzCDFSAEkSAG8SAF9skgIRZxLhZgAmEnlaStoSPpIAbToAdhASADl+YXHcHABdZhzG4WaxVoAVEl91kM3hGxJTVILSTbG6NoD7Osz6DosSMa6aHpp+xZJ+BwLnKDkSTbbXYkfaHY7OU6PdqGIIg65AA==)

**Diagnostic Id**: CS8764: Nullability of return type doesn't match overridden member (possibly because of nullability attributes)

**Expected Behavior**: There should be a CS8764 warning on overriding `[return: NotNull] T A.M` with `T B.M`. The override effectively loses `[NotNull]` annotation making the behavior of override method less safe than the base member.

**Actual Behavior**: No warnings at all in the program above. The program crashes at runtime with a `NullReferenceException`
